fendermint/vm/interpreter/src/fvm/executions.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::fvm::state::FvmExecState;
+use crate::types::*;
+use anyhow::Context;
+use fendermint_vm_actor_interface::system as system_actor;
+use fendermint_vm_actor_interface::{chainmetadata, cron, system};
+use fendermint_vm_message::signed::SignedMessage;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::RawBytes;
+use fvm_shared::address::Address;
+use fvm_shared::BLOCK_GAS_LIMIT;
+use ipc_observability::{emit, measure_time};
+
+use crate::fvm::observe::{MsgExec, MsgExecPurpose};
+
+use crate::fvm::FvmMessage;
+
+const GAS_LIMIT: u64 = BLOCK_GAS_LIMIT * 10000;
+
+/// Helper to build and execute an implicit system message.
+/// It uses the default values for the other fields not passed.
+fn execute_implicit_message<DB: Blockstore + Clone + 'static + Send + Sync>(
+    state: &mut FvmExecState<DB>,
+    from: Address,
+    to: Address,
+    sequence: u64,
+    gas_limit: u64,
+    method_num: u64,
+    params: RawBytes,
+) -> anyhow::Result<AppliedMessage> {
+    let msg = FvmMessage {
+        from,
+        to,
+        sequence,
+        gas_limit,
+        method_num,
+        params,
+        value: Default::default(),
+        version: Default::default(),
+        gas_fee_cap: Default::default(),
+        gas_premium: Default::default(),
+    };
+
+    let (apply_ret, emitters) = state.execute_implicit(msg)?;
+    if let Some(err) = apply_ret.failure_info {
+        anyhow::bail!("failed to apply system message: {}", err);
+    }
+    Ok(AppliedMessage {
+        apply_ret,
+        emitters,
+        from,
+        to,
+        method_num,
+        gas_limit,
+    })
+}
+
+/// Executes a signed message and returns the applied message.
+pub async fn execute_signed_message<DB: Blockstore + Clone + 'static + Send + Sync>(
+    state: &mut FvmExecState<DB>,
+    msg: SignedMessage,
+) -> anyhow::Result<AppliedMessage> {
+    let msg = msg.into_message();
+
+    // Select execution path based on sender.
+    let (apply_ret, emitters, execution_time) = if msg.from == system_actor::SYSTEM_ACTOR_ADDR {
```

Comment:
I think a signed message can never have `msg.from == SYSTEM_ACTOR_ADDR`. That means something is wrong, we should remove this branch.

Reply:
Yes, this is a good point. I kept there as legacy, but I agree it could be removed.